### DISCUSSION TITLE
[Question] Any way to add query parameters and body parameters in the same request?

### DISCUSSION THREAD

**Discussion by gperdomor** [Alamofire/Alamofire#3862] [Ideas]
Hi... I wanna send a request with query and body parameters...

Alamofire support this feature or i need to build the query parameters manually like this:

```swift

Alamofire.request("MY_URL?access_token=\(token)&param2=\(param2)", method: .post ...)
```

**Comment by jshier** [comment]
It's not built into the top level APIs but you should be able to pass a `URLRequest` through two different encodings with no issue. I'd recommend encapsulating such logic inside a [router](https://github.com/Alamofire/Alamofire#routing-requests) as described in our documentation.

**Comment by jj10133** [comment]
Hi @jshier I recently found this library however I have the same issue that I have a post request with query parameter and body? How could I achieve this? I look at the wiki examples but couldn't find, looked into QueryString for URLEncoded but still...

**Comment by jshier** [comment]
Things haven't fundamentally changed since my original reply. Alamofire doesn't offer anything built in to do this, but you should be able to compose our various `ParameterEncoder` implementations inside a `URLRequestConvertible` implementation to encode parameters wherever you want. You could even do so generically.

```swift
struct QueryAndBodyParameters<Query: Encodable, Body: Encodable>: URLRequestConvertible {
  let query: Query
  let body: Body

  func asURLRequest() throws -> URLRequest {
    var request = ...
    let queryEncoder = URLFormEncodedParameterEncoder(destination: .queryString)
    try queryEncoder.encode(query, into: &request)
    let bodyEncoder = JSONParameterEncoder()
    try bodyEncoder.encode(body, into: &request)

    return request
  }
}
```

Something like that can provide the basic structure you need.

I'll convert this to a discussion in case other have any ideas.

**Reply by jshier** [reply]
If you'd like an overall introduction to Alamofire, I suggest you read our [Usage guide](https://github.com/Alamofire/Alamofire/blob/master/Documentation/Usage.md#using-alamofire). You can find even more detail in our [Advanced usage guide](https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#advanced-usage).
